src/vision/status_quo/barbara_carefully_dismisses_embedded_future.md

```diff
@@ -246,6 +246,69 @@ instead. Here are some ways in which these APIs are lighter weight than a
       community and ecosystem.
     * [Niklaus] would really have struggled. If he asked for help, he probably
       would've gotten conflicting advice from the community.
+* **`Future` has a lot of features that Barbara's traits don't have -- aren't
+    those worth the cost?**
+    * `Future` has many additional features that are nice-to-have:
+        1. `Future` works smoothly in a multithreaded environment. Futures can
+           be `Send` and/or `Sync`, and do not need to have interior mutability,
+           which avoids the need for internal locking.
+           * Manipulating arbitrary Rust types without locking allows `async fn`
+             to be efficient.
+        1. Futures can be spawned and dropped in a dynamic manner: an executor
+           that supports dynamic allocation can manage an arbitrary number of
+           futures at runtime, and futures may easily be dropped to stop their
+           execution.
+           * Dropping a future will also drop futures it owns, conveniently
+             providing good cancellation semantics.
+           * A future that creates other futures (e.g. an `async fn` that calls
+             other `async fn`s) can be spawned with only a single memory
+             allocation, whereas callback-based approaches need to allocate for
+             each asynchronous component.
+        1. Community and ecosystem support. This isn't a feature of `Future` per
+           se, but the Rust language has special support for `Future`
+           (`async`/`await`) and practically the entire async Rust ecosystem is
+           based on `Future`. The ability to use existing async crates is a very
+           strong reason to use `Future` over any alternative async abstraction.
+    * However, the code size impact of `Future` is a deal-breaker, and no number
+      of nice-to-have features can outweigh a deal-breaker. Barbara's traits
+      have every feature she *needs*.
+    * Using `Future` saves developer time relative to building your own async
+      abstractions. Developers can use the time they saved to minimize code size
+      elsewhere in the project. In some cases, this may result in a net decrease
+      in code size for the same total effort. However, code size reduction
+      efforts have diminishing returns, so projects that expect to optimize code
+      size regardless likely won't find the tradeoff beneficial.
+* **Is the code size impact of `Future` fundamental, or can the design be
```

Comment:
The question isn't directly addressed in the answer (and maybe more design work needs to happen before it can be answered), but it doesn't feel like this is a *fundamental* tradeoff. If `Waker` could be tweaked to allow the code doing the waking to provide relevant context of who they are to the `Waker` perhaps this could be addressed. Thoughts?

Reply:
The graph of possible execution flows for a future that is waiting on two events merges and diverges:

```
+-------+    +-------+
|Event A|    |Event B|
+-------+    +-------+
      |        |
      V        V
    +------------+
    |Waker::wake |
    +------------+
          |
          V
    +------------+
    |Future::poll|
    +------------+
       |      |
       V      V
+--------+  +--------+
|Handle A|  |Handle B|
+--------+  +--------+
```

As long as this merge exists, `Future::poll` will need to do work to determine what work triggered the poll. Callback-based systems transmit this implicitly, because the execution paths never merge.

I'm currently unsure whether it's possible to avoid the execution path merge without breaking backwards compatibility. It may be possible to add a `poll_with_context<T>()` method to `Future`, with a corresponding `wake_with_context<T>()` method on `Waker`, so that code paths for different events remain distinct.

Reply:
I feel like the existing answer suffices. I'm not sure exactly what would be worth saying to make it more detailed.